src/web/headers.rs

```diff
@@ -1,10 +1,31 @@
+use super::encode_path_for_uri;
+use anyhow::Result;
 use axum::{
     headers::{Header, HeaderName, HeaderValue},
-    http::uri::Uri,
+    http::uri::{PathAndQuery, Uri},
 };
+use serde::Serialize;
 
 /// simplified typed header for a `Link rel=canonical` header in the response.
-pub struct CanonicalUrl(pub Uri);
+/// Only takes the path to be used, url-encodes it and attaches domain & schema to it.
+#[derive(Debug, Clone)]
+pub struct CanonicalUrl(PathAndQuery);
+
+impl CanonicalUrl {
+    pub fn from_path<P: AsRef<str>>(path: P) -> Result<Self> {
+        Ok(Self(encode_path_for_uri(path)?))
+    }
+
+    fn build_full_uri(&self) -> Uri {
+        Uri::builder()
+            .scheme("https")
+            .authority("docs.rs")
```

Comment:
Why are we now using absolute redirects? We were using relative redirects before, right? (I saw one of the tests changed.)